src/main.rs

```diff
@@ -124,6 +128,63 @@ fn respond_handle_error(result: Result<ResponseFuture, Box<dyn Error>>) -> Respo
     }
 }
 
+fn get_static(req: Request<Body>) -> Result<ResponseFuture, Box<dyn Error>> {
+    let relative_asset_path = &req.uri().path()[STATIC_ASSET_PATH.len()..];
+    trace!("Loading asset located at relative path: {}", relative_asset_path);
+
+    let asset_filesystem_path: String = format!(
+        "{}/{}",
+        STATIC_ASSET_FILESYSTEM_ROOT,
+        relative_asset_path
+    );
+    trace!("Computed non-canonicalized filesystem path: {}", asset_filesystem_path);
+
+    let asset_canonicalized_path = fs::canonicalize(asset_filesystem_path)?;
+    trace!("Canonicalized filesystem path: {}", asset_canonicalized_path.to_str().unwrap());
+
+    let pwd = env::current_dir()?;
+
+    let absolute_begins_with_path = format!("{}/{}", pwd.to_str().unwrap(), STATIC_ASSET_FILESYSTEM_ROOT);
+
+    if(
+        !asset_canonicalized_path
+            .to_str()
+            .unwrap()
+            .starts_with(&absolute_begins_with_path)
+    ) {
+        // Looks like someone tried path traversal.
+        // just return a 404 and forget about it.
+        return Ok(Box::new(future::ok(
+            Response::builder()
+                .status(404)
```

Comment:
This would be better done by expanding the error type and `handle_error` functions to provide a human-readable message (if applicable) and response code.